Migration of meta-qcom (and merge with meta-qcom-hwe)

ndechesne opened the pull request:
Related to: https://github.com/Linaro/meta-qcom/issues/680

This is probably impossible to properly review in Github. The key commits to review are the most recent commits in the PR branch. 

First I am doing a 'non trivial' merge of meta-qcom (master) into meta-qcom-hwe (main) using "git merge --allow-unrelated-histories". Most of the merge is trivial with no conflicts, e.g. merging recipes, conf files, ... from both layers, just fine. The main conflicts were with README.md and conf/layer.conf. For both of these files, I have done a manual merge/edit so that they look right to me. These files need to be reviewed carefully. 

The next 2 commits are changes in the Github workflows. What I am trying to do with the workflows is:
* For all stable up to styhead, they predate the migration, so Linaro will continue to maintain them, and the CI should be based on existing Linaro CI
* For master and all future stable branches, we are moving to Qualcomm CI

Note: the Linaro CI is not expected to work for now, since we will need to setup the secrets/token.

The result of the merge represents a very decent history of meta-qcom with ~1650 commits from ~50 developers over the last 10 years.

In terms of process, we eventually want to migrate meta-qcom into github.com/quic-yocto org. We can decide to merge this PR (when it's tested/ready) and continue the development of meta-qcom-hwe (main branch) here after the initial merge. When the repo is migrated, we will just need to do a final merge of the meta-qcom-hwe main branch into meta-qcom repo.

ndechesne commented:
Submitted the following fixes to meta-qcom, which should fix the patchreview and yocto-check-layer issues observed in our CI.

https://github.com/Linaro/meta-qcom/pull/683
https://github.com/Linaro/meta-qcom/pull/684

ndechesne commented:
I ran 2 successive builds with meta-qcom-hwe/main first, and then with this PR. buildhistory reports no change in between, which is a good sign!

commit 6b7b722169b4b87ef8e8e58fa8cef93d06c51574 (HEAD -> master)
Author: buildhistory <buildhistory@poky>
Date:   Thu Nov 28 16:21:54 2024 +0000

    No changes: Build 20241128162147 of poky 5.1 for machine qcs6490-rb3gen2-core-kit on cfd6913e01e5

    cmd: bitbake -c build esp-qcom-image core-image-minimal

    result: succeeded

    metadata revisions:
    repo              = pr/92:5a3e2fa7302d508dcaba43fc5d6e44d208342d2f
    meta              = master:7434b5428971576c6c1422d0a385084936ee9768
    meta-poky         = master:7434b5428971576c6c1422d0a385084936ee9768

diff --git a/metadata-revs b/metadata-revs
index bef71f4..feb2ca7 100644
--- a/metadata-revs
+++ b/metadata-revs
@@ -1,4 +1,3 @@
-repo              = HEAD:bdb293260d5ccae850364d1f38aa77509940b61f
-meta-qcom         = master:72264dd4c526ce7606425d7badbbde3a350e4226
+repo              = pr/92:5a3e2fa7302d508dcaba43fc5d6e44d208342d2f
 meta              = master:7434b5428971576c6c1422d0a385084936ee9768
 meta-poky         = master:7434b5428971576c6c1422d0a385084936ee9768

lumag commented:
Thanks for the fixes!
Just for my understanding, could you please list the configurations that are used for the testing? Do you test qcom-armv8a/v7a with the merged branch?

ndechesne commented:
For now, I did not change any of the CI targets, so it's only what we had in meta-qcom-hwe, e.g. https://github.com/quic-yocto/meta-qcom-hwe/blob/main/.github/workflows/build-yocto.yml, so poky+this layer for the 3 non generic platforms we have. I would like to first merge without adding more CI, then we can fix up / improve the CI before the final migration.

lumag commented:
That means that we have to manually validate the merged branch with generic configurations, thanks for the note.

ndechesne commented:
Well, perhaps we should add them in the branch, you might be right.. any reason not to add the generic machines in the default CI, @ricardosalveti ?

ricardosalveti commented:
> Well, perhaps we should add them in the branch, you might be right.. any reason not to add the generic machines in the default CI, @ricardosalveti ?

I would prefer us moving towards genericarm64, but until that is done and for the purposes of this migration, I don't see an issue on having the current meta-qcom generic targets built by our CI.

lumag commented:
@ricardosalveti `genericarm64` is a generic machine targeting SystemReady devices. Unless one replaces ABL with U-Boot devices before RB3gen2 are not SR-compatible. As such in my opinion it doesn't really make sense to support those in the `genericarm64` machine.

lumag commented:
So, generally speaking: LGTM. I've opened https://github.com/Linaro/meta-qcom/pull/685, but that can also be merged afterwards. The only remaining item from my side is testing `qcom-armv8a` / `qcom-armv7a` in the CI.

lumag commented:
@ndechesne are there any remaining blockers? It would be nice to sort it out and complete migration.

ndechesne commented:
@lumag I am traveling this week, and I don't have a lot of free cycles to finish. I am hoping we can make more progress next week..

I found some issues with the CI patches I made. I think we need more changes there. Since the 'master' branch is going to run on the QCOM CI, we need to move all Linaro CI outside of the master branch. So all branches up to styhead will be 'linaro' stuff, and master and future releases branches will be on QCOM CI. I think we should move all Linaro workflow either to styhead branch or perhaps scarthgap branch since it's the 'last' LTS branch on Linaro CI. 

e.g. have all the workflow yaml files in scarthgap, and update all other branches to point to scarthgap workflows. Currently scarthgap and styhead have a copy of the workflows, kirkstone refers to workflow from 'master' branch (with: uses: Linaro/meta-qcom/.github/workflows/build-template.yml@master). and older branches (honister and before) are broken, as they refer to workflow in ndechesne/meta-qcom ( uses: ndechesne/meta-qcom/.github/workflows/build-template.yml@master).. 

So do you want to get the Linaro CI updated/fixed in all branches other than master first? 

I am planning to the Linaro CI workflow from this PR (when I get back to it)

lumag commented:
I completely forgot about that. I will take a look at the CI for other branches

koenkooi commented:
FWIW, a kas based build for qcs6490-rb3gen2-core-kit using this PR builds without issues.

lumag commented:
@ndechesne I think I've fixed all outstanding CI issues.

igoropaniuk commented:
Taking into consideration that there are currently 113 commits in `meta-qcom-hwe` and it's not currently used in production, is there any reasons why the decision was made to merge it this way and not the other way around: fork `meta-qcom` first and then merge `meta-qcom-hwe` into it? This would definitely have required less effort in merging and follow-up testing, and as a result we would end up with 1:1 replication of `meta-qcom` git history

Another concern is that as for `meta-qcom-hwe` "rebase and merge" strategy was chosen from the very beginning, and pulling all merge commits from `meta-qcom` IMO might spoil existing git history in `meta-qcom-hwe` (I know it's just a matter of preference). Have we considered get rid of merge commits (I know that it's a huge amount work, but anyway)?

ndechesne commented:
> Taking into consideration that there are currently 113 commits in `meta-qcom-hwe` and it's not currently used in production, is there any reasons why the decision was made to merge it this way and not the other way around: fork `meta-qcom` first and then merge `meta-qcom-hwe` into it? This would definitely have required less effort in merging and follow-up testing, and as a result we would end up with 1:1 replication of `meta-qcom` git history
> 

git merge A from B, or git merge B from A. I don't see how it makes the merge any different. The same merge conflicts would exist. And in fact, there really was no difficult merge at all. It was fairly trivial to do the merge, the way I did it. 

Since we are merging the 2 branches, we still keep the history of both branches, so the history of meta-qcom is replicated. Can you please elaborate? I am open to doing things in a different way if that makes anything simpler.

> Another concern is that as for `meta-qcom-hwe` "rebase and merge" strategy was chosen from the very beginning, and pulling all merge commits from `meta-qcom` IMO might "spoil" existing git history in `meta-qcom-hwe` (I know it's just a matter of preference, but anyway there will be some chaos). Have we considered getting rid of merge commits (I know that it's a huge amount work, but just curious if it's feasible)?

That is correct. the history is going to be a bit of mess, but moving forward we will adopt rebase and merge again. As long as we kept the 2 branches history in the merge, I think it's fine.

igoropaniuk commented:
@ndechesne 
> Since we are merging the 2 branches, we still keep the history of both branches, so the history of meta-qcom is replicated. Can you please elaborate? I am open to doing things in a different way if that makes anything simpler.

I got confused as I was thinking about rebase for some reason when I wrote "1:1 replication of meta-qcom git history", however you explicitly mentioned "merge" in the PR description, so sorry for the nohistoryise.
My point was that having PR with 112 commits would look more lightweight and less scary :) , but I do agree that in the end we would get the same git tree and it's just a matter of preference

koenkooi commented:
When using 'kas build' for the rbd3, this doesn't seem to generate a qcomflash tarball anymore.

igoropaniuk commented:
@koenkooi I think that's because it doesn't include this commit:
```
commit d6358019cce14f21e95a6a31cec28b134e2fa0d7
Author: Ricardo Salveti <ricardo.salveti@oss.qualcomm.com>
Date:   Thu Nov 14 19:22:20 2024 -0300

    image_types_qcom: add support for qcomflash IMAGE_FSTYPE
    
    Add support for a new IMAGE_FSTYPE which creates a single tarball
    containing all the relevant files to perform a full clean flash,
    including partition metadata, boot firmware, ESP partition and the
    rootfs.
    
    Signed-off-by: Ricardo Salveti <ricardo.salveti@oss.qualcomm.com>

```

ndechesne commented:
I updated the PR. What's changed:
1. I did the merge again with the most recent commits from meta-qcom and meta-qcom-hwe
2. completely disabled (removed the workflow) the Linaro CI in main
3. added meta-qcom generic machines in CI

lumag commented:
Nice! Ack from my side.

ndechesne commented:
@ricardosalveti when we decide to merge this PR, we will need to do a merge commit, it's an exception to our process, but i believe that a rebase/squash will break everything..

ndechesne commented:
I pushed a fix in meta-qcom that should address the yocto-check-layer error. It was a problem in the original Linaro/meta-qcom which we found during the merge.

sbanerjee-quic reviewed (approved):
LGTM

ricardosalveti reviewed (approved):
LGTM

quaresmajose reviewed (approved):
LGTM
